src/jaxsim/rbda/contacts/visco_elastic.py

```diff
@@ -0,0 +1,1050 @@
+
+        I_nc = jnp.eye(v_t0.flatten().size)
+        O_nc = jnp.zeros(shape=(p_t0.flatten().size, p_t0.flatten().size))
+
+        # Pack the A matrix.
+        A = jnp.vstack(
+            [
+                jnp.hstack([O_nc, I_nc, O_nc]),
+                ψ @ jnp.split(A_sc, 2, axis=0)[0],
+                jnp.split(A_sc, 2, axis=0)[1],
+            ]
+        )
+
+        # Short names for few variables.
+        ν = BW_ν
+        J = jnp.vstack(CW_Jl_WC)
+        J̇ = jnp.vstack(CW_J̇l_WC)
+
+        # Compute the free system acceleration components.
+        with (
+            data.switch_velocity_representation(jaxsim.VelRepr.Mixed),
+            references.switch_velocity_representation(jaxsim.VelRepr.Mixed),
+        ):
+
+            BW_v̇_free_WB, s̈_free = js.ode.system_acceleration(
+                model=model,
+                data=data,
+                joint_force_references=references.joint_force_references(model=model),
+                link_forces=references.link_forces(model=model, data=data),
+            )
+
+        # Pack the free system acceleration in mixed representation.
+        ν̇_free = jnp.hstack([BW_v̇_free_WB, s̈_free])
+
+        # Compute the acceleration of collidable points.
+        # This is the true derivative of ṗ only in mixed representation.
+        p̈ = J @ ν̇_free + J̇ @ ν
+
+        # Pack the b array.
+        b = jnp.hstack(
+            [
+                jnp.zeros_like(p_t0.flatten()),
+                p̈ + ψ @ jnp.split(b_sc, indices_or_sections=2)[0],
+                jnp.split(b_sc, indices_or_sections=2)[1],
+            ]
+        )
+
+        return A, b, A_sc, b_sc
+
+    @staticmethod
+    @functools.partial(jax.jit, static_argnames=("terrain",))
+    def _linearize_contact_model(
+        position: jtp.VectorLike,
+        velocity: jtp.VectorLike,
+        tangential_deformation: jtp.VectorLike,
+        parameters: ViscoElasticContactsParams,
+        terrain: Terrain,
+    ) -> tuple[jtp.Matrix, jtp.Vector]:
+        """"""
+
+        # Initialize the state at which the model is linearized.
+        p0 = jnp.array(position, dtype=float).squeeze()
+        v0 = jnp.array(velocity, dtype=float).squeeze()
+        m0 = jnp.array(tangential_deformation, dtype=float).squeeze()
+
+        # ============
+        # Compute A_sc
+        # ============
+
+        compute_contact_force_non_linear_model = functools.partial(
+            ViscoElasticContacts.compute_contact_force_non_linear_model,
+            parameters=parameters,
+            terrain=terrain,
+        )
+
+        # Compute with AD the Jacobians of CW_fl w.r.t. the inputs.
+        df_dp_fun, df_dv_fun, df_dm_fun = (
+            jax.jacrev(
+                lambda p0, v0, m0: compute_contact_force_non_linear_model(
+                    position=p0, velocity=v0, tangential_deformation=m0
+                )[0],
+                argnums=num,
+            )
+            for num in (0, 1, 2)
+        )
```

Comment:
Can this work as well?
```suggestion
        df_dp_fun, df_dv_fun, df_dm_fun = jax.vmap(
            lambda num: jax.jacrev(
                lambda p0, v0, m0: compute_contact_force_non_linear_model(
                    position=p0, velocity=v0, tangential_deformation=m0
                )[0],
                argnums=num,
            ),
            num=jnp.arange(3),
        )
```

Reply:
Good question, I have no clue. Let me check.

Reply:
These functions do not return the computation of the jacobian, they just return the function to compute the jacobian. A function is not a valid JAX type, therefore it cannot be vmapped:

```
TypeError: Output from batched function <function ViscoElasticContacts._linearize_contact_model.<locals>.<lambda>.<locals>.<lambda> at 0x7042f6b44ae0> with type <class 'function'> is not a valid JAX type
```

Reply:
You're right! Maybe in the future we can this of a refactor to make the computation happen in a self-contained vmap-able function. Thanks for investigating on this!

Reply:
Note that this function is already vmapped on the contact points, and this means that these jacobians are computed in parallel for all contact points.

Reply:
Yes! Still, that `for` loop might introduce a small increase of the compilation time, but I believe we can ignore it for the time being